Key/Value Formats

quodlibetor opened the pull request:
Interesting parts of this change:

* Two new types: CreateSourceFormat/SourceDataEncoding, which wrap the
  encoding definition with something that can specify Key/Value or just
  Value encoding. Currently only kafka supports this.
  * The Upsert syntax (UPSERT FORMAT ...) is still supported, but it
    emits a warning and gets normalized to KEY FORMAT / VALUE FORMAT internally
* KeyValueDecoder, which takes a pair of boxed DecoderStates to handle
  key/value decoding.
* All of the upsert key/value rewriting is now internal to the upsert
  code, it no longer affects anything externally.
* We can now handle arbitrary encodings for keys/values with a unified
  framework, everything should just work when we add new encodings in
  the future.
* The legacy upsert key format embedding is still supported with a
  WARN-level log, but the new syntax is also supported (see the new
  kafka-upsert-new TD test file.)

quodlibetor commented on src/dataflow/src/decode/mod.rs:
```diff
@@ -441,6 +385,63 @@ fn decode_cdcv2<G: Scope<Timestamp = Timestamp>>(
+        SourceDataEncoding::KeyValue { key, value } => {
+            let op_name = format!("Key{}Value{}Decode", key.op_name(), value.op_name());
+
+            // XXX: if the value is CDCv2 should we bail here, since that is the only thing that can
+            // return a token in decode_values_single?
+            let decoder = get_decoder(key, value, &envelope).expect("decoder previously validated");
```
This should be resolved before merging

quodlibetor replied:
done, now.

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -1648,13 +1643,64 @@ impl<'a> Parser<'a> {
+                    } else {
+                        self.error(
+                            self.index,
+                            "Key/Value format specified multiple times".into(),
```
Let's say something like `ENVELOPE UPSERT FORMAT conflicts with earlier KEY FORMAT specification` to make it more clear which clauses are in conflict!

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -1648,13 +1643,64 @@ impl<'a> Parser<'a> {
+                            "Upsert requires a value format"
+                        );
+                    };
+                }
```
I don't think it makes sense to do this normalization in the parser. (I.e., this last `format.is_simple()` branch.) Normalizing the `ENVELOPE UPSERT` into the new `KEY FORMAT` seems ok because it will facilitate the migration, but this normalization looks like it will stick around a while and it really seems like semantic analysis to me. Like, the planner should just do this interpretation of `CreateSourceFormat::Bare`.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1083,6 +940,182 @@ pub fn plan_create_sources(
+        Format::Json => unsupported!("JSON sources"),
+        Format::Text => DataEncoding::Text,
+    }))
+}
```
I can see why you wanted to split these functions out but it's real hard to see what changed as a result :/

quodlibetor replied:
I'll revert this.

umanwizard replied:
was this addressed?

quodlibetor replied:
whoops, no, it is now though.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -922,28 +796,15 @@ pub fn plan_create_source(
+        sql_parser::ast::Envelope::Upsert => match connector {
+            Connector::Kafka { .. } => match format {
+                CreateSourceFormat::KeyValue { .. } => SourceEnvelope::Upsert,
+                // this is validated to become keyvalue in purify, where we can talk to CSR
```
I'm not sure I follow this comment.

umanwizard replied:
I think the point is that logically upsert sources need both a key and value format, which is why it's confusing to accept `Bare` here.

However, to match historical practice, we allow _only_ a value format if and only if it happens to be Avro + CSR. In that case, we then get the corresponding key format from CSR as well.

(This could indeed use some expansion to be comprehensible to people who don't understand the historical context).

Also now I can't find the code that does what I'm talking about so maybe I'm just totally confused and @quodlibetor can correct me.

quodlibetor replied:
oh, you're correct brennan but the comment is slightly wrong: it still happens in two parts:

* We determine that there is a key/value *schema* in purify
* We convert from a single term to a keyvalue term when we rewrite from `Format` -> `Encoding` later

There is a small comment on [line 949 of `plan/statement/ddl.rs`](https://github.com/MaterializeInc/materialize/pull/6487/files#diff-c840a7fbb6a642d6cbf71dddd92ec1b6dfa86310022ead5673e8fa509ba73dc4R949) (inside `get_encoding`) that mentions the fact that this is ... bad. IMO it _should_ be in purify, but as mentioned this PR is already too large.

benesch reviewed (commented):
I took a first skim through this and my first thought is: whoa this is large! I'm probably starting to sound like a broken record, but with something this large and fiddly it's hard to give a quality review. I would love if we could move towards a model of smaller, more frequent PRs by default. I know that puts an additional burden on you so happy to help however I can (e.g., with speedier reviews). On the bright side, it's all passing tests, it seems? So that's exciting!

I made it through the parser stuff so far. Generally looks good, but a few notes on where it looks like some semantic analysis that ought to be performed in the planner has crept in. 

I'm not very familiar with the purify/`CREATE SOURCE` code these days. If @umanwizard can make it through the diff in the current form that's great. Otherwise I will need some more handholding.

quodlibetor commented:
> whoa this is large!

I wholly agree with this! As I mentioned in my standup several times, I lost 3 or 4 days trying to get this smaller. There are a couple one offs that I could move around into their own PRs (like that function getting moved) but unfortunately this PR is trying to get rid of a bunch of really distantly coupled code. Everything that I tried to do to keep this smaller either wouldn't compile or made for nonsensical PRs. I'm very happy to get on a video call with you @benesch or @umanwizard if it'll help.

quodlibetor commented on src/sql-parser/src/parser.rs:
```diff
@@ -1648,13 +1643,64 @@ impl<'a> Parser<'a> {
-            Some(self.parse_format()?)
+        // legacy upsert format syntax allows setting the key format after the keyword UPSERT, so we
+        // may mutate this variable in the next block
+        let mut format = if matches!(connector, Connector::Kafka { .. }) {
```
Do we verify things in the planner or in purify? Happy to move it wherever.

benesch replied:
Either is fine; between purify/the planner, wherever is most convenient/necessary.

quodlibetor replied:
moved it to plan

quodlibetor commented on src/sql-parser/src/parser.rs:
```diff
@@ -1792,7 +1839,13 @@ impl<'a> Parser<'a> {
+                // * `KEY FORMAT` means there is no key, we'll parse a KeyValueFormat later
+                let is_key_format = (matches!(self.peek_keyword(), Some(KEY))
+                    && matches!(self.peek_nth_token(1), Some(Token::Keyword(FORMAT))));
+                let key = if !is_key_format && self.peek_keyword() == Some(KEY) {
```
Yeah I agree, I wrote it like this to make it more obvious what the intent was but if it's not clear that that was the intent then my good intentions lead me astray.

cuongdo commented:
@umanwizard your wizardly knowledge of sources would be useful in reviewing this code

umanwizard commented:
> @umanwizard your wizardly knowledge of sources would be useful in reviewing this code

I'm definitely reviewing this and have been closely involved in the discussions!

quodlibetor commented:
I've responded to all code review comments.

cuongdo commented:
@umanwizard do you have further comments regarding this PR?

umanwizard commented:
I've been out of the office since the most recent update was posted -- reviewing today.

umanwizard reviewed (commented):
Looks mostly sane, just a few comments.

quodlibetor reviewed (commented):
Updated responding to @umanwizard's review

umanwizard commented:
the .td failure seems real

quodlibetor commented:
it does indeed.

quodlibetor commented:
@umanwizard I reverted that internal error and added a comment that would explain to me why it can't be applied, which I forgot again because I am very much looking forward to trashing that code.

benesch reviewed (approved):
The parser stuff LGTM! I didn't look at any of the rest.
